core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -114,6 +117,10 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
             return Result.empty();
         }
 
+        if (getJoinPushdownMode(context.getSession()) == JoinPushdownMode.AUTOMATIC && !shouldProceedWithPushDown(joinNode, context)) {
```

Comment:
I think the `getJoinPushdownMode` should be consulted inside `shouldProceedWithPushDown`
(or you'd want to rename the method to indicate it's appropriate for "automatic" mode only)

Reply:
Renamed the method to `skipJoinPushdownBasedOnCost` (reversing true/false return value semantics), and moved `getJoinPushdownMode(context.getSession()) == JoinPushdownMode.AUTOMATIC` inside.